src/ring_theory/roots_of_unity.lean

```diff
@@ -153,6 +153,32 @@ calc  fintype.card (roots_of_unity k R)
 ... = (nth_roots k (1 : R)).card                  : multiset.card_attach
 ... ≤ k                                           : card_nth_roots k 1
 
+variables {k R}
+
+/-- Restrict a ring homomorphism between integral domains to the nth roots of unity -/
+def ring_hom.restrict_roots_of_unity (σ : R →+* S) (n : ℕ+) :
+  roots_of_unity n R →* roots_of_unity n S :=
+let h : ∀ ξ : roots_of_unity n R, (σ ξ) ^ (n : ℕ) = 1 := λ ξ, by
+{ change (σ (ξ : units R)) ^ (n : ℕ) = 1,
+  rw [←σ.map_pow, ←units.coe_pow, show ((ξ : units R) ^ (n : ℕ) = 1), from ξ.2,
+      units.coe_one, σ.map_one] } in
+{ to_fun := λ ξ, ⟨is_unit.unit (is_unit_of_pow_eq_one (σ ξ) n (h ξ) n.2),
```

Comment:
After #6658 you can use invertible_of_pow_eq_one here to avoid needing the non-computable `is_unit.unit`.

Reply:
Are you suggesting to use `unit_of_invertible` and `invertible_of_pow_eq_one`? The sounds reasonable, although currently `unit_of_invertible` uses `[invertible a]`, so I would have to do something like `@unit_of_invertible _ _ _ (invertible_of_pow_eq_one _ _ (h ξ) n.2)`, which works, but is a bit painful.

Reply:
`let i := invertible_of_pow_eq_one _ _ (h ξ) n.2 in by exactI unit_of_invertible (σ ξ)` is perhaps sligghtly less painful, but I see your point.

Reply:
However, if it can avoid the `is_unit.unit_spec` stuff later on, then it might be worth it.

Reply:
I merged your PR into mine, and I think that it makes things less painful.

Reply:
Although it seems to have caused a cyclic import that I don't understand